spec/GraphQLOverHTTP.md

```diff
@@ -112,72 +178,152 @@ http://example.com/product/graphql
 
-## Content Types
+A server MAY support requests encoded with and/or accepting other media types.
 
-The following are the officially recognized GraphQL content types to designate encoding JSON over HTTP.
+If the client does not supply a `Content-Type` header, the server SHOULD treat
+the request as if it had `Content-Type: application/json`.
 
-| Name  | Description |
-| ------------- | ------------- |
-| `application/graphql+json` | Required |
-| `application/json` | To support legacy clients |
+If the client does not supply an `Accept` header, the server SHOULD treat the
+request as if it had `Accept: application/json`.
 
-A server MUST support requests from clients with HTTP header `Content-Type: application/graphql+json` indicating that the body of the request is a JSON document with a GraphQL request. A server must indicate the content type of the response with a `Content-Type` header. This default value should be `Content-Type: application/graphql+json` indicating that the response is a valid JSON document [conforming to the GraphQL spec](https://spec.graphql.org/October2021/#sec-Response-Format).
+Note: These defaults are in place to maintain compatibility with legacy clients.
 
-A server MAY support requests from clients with other content types.
+Note: Prior to this specification, the media type `application/json` was in wide
+use for both the HTTP request body type and the HTTP response payload type.
+Unfortunately this causes a number of issues, not least that it means clients
+cannot trust responses from the server that do not use an HTTP 2xx status code
+(since these replies may come from non-compliant HTTP servers or proxies
+somewhere in the request chain). For this reason, this specification introduces
+the `application/graphql+json` media type; however, to give existing servers
+time to move over, it is not required to be supported until 1st January 2025.
 
-A client MUST handle receiving responses with HTTP header `Content-Type: application/graphql+json` since any compliant server must support this type.
+### Legacy watershed
 
-This header MAY include encoding information (e.g. `Content-Type: application/graphql+json; charset=utf-8`)
+From 1st January 2025 (`2025-01-01T00:00:00Z`), a server MUST support requests
+encoded with the `application/graphql+json` media type (as indicated by the
+`Content-Type` header).
+
+From 1st January 2025 (`2025-01-01T00:00:00Z`), a server MUST support requests
+which accept the `application/graphql+json` media type (as indicated by the
+`Accept` header).
 
 # Request
 
-A server MUST accept POST requests, and MAY accept other HTTP methods, such as GET.
+A server MUST accept POST requests, and MAY accept other HTTP methods, such as
+GET.
 
 ## Request Parameters
 
-A request for execution should contain the following request parameters:
+A {GraphQL-over-HTTP request} is formed of the following parameters:
+
+- {query} - A Document containing GraphQL Operations and Fragments to execute.
```

Comment:
This also should be optional to allow for persisted queries, but I can do a separate PR on this outlining the use-case. I original added the `{extensions}` for this so the persisted query id can now go into that map.

Reply:
I don't think that a server that only allows persisted operations is a truly GraphQL-over-HTTP compliant server. The literal intent of an operation allowlist is to _prevent_ interoperability, and that's okay!

Reply:
> The Apollo implementation of automatic persisted queries works by leaving out query in the particular requests [...] If we want to try to make the Apollo APQ protocol not a violation of this spec, we could make query optional and have some sort of note stating that when query is left out, the server must have some sort of extensions-based mechanism for deriving the Document.

I think it's okay for Automatic Persisted Queries to not be an official GraphQL request - the point of the specification is to define a common set of rules that means a compliant client from any vendor can speak to a compliant server from any vendor. APQ is not part of that flow, it's a specific behavior that the client/server can negotiate as an extension, separate from this specification.

> Does the spec actually explain anywhere what to do if the query is not provided? I suppose for GETs this falls into the "Server URLs which enable GraphQL requests MAY also be used for other purposes" thing to some degree.

"A {GraphQL-over-HTTP request} is formed of the following parameters" implies that missing "query" means that it's not a "GraphQL over HTTP request"; then we have "When a server receives a {GraphQL-over-HTTP request}, it must return a well‐formed response." but it's not a GraphQL over HTTP request, so it's up to the server how to handle it, either by classifying it as "other purposes" (in this case serving GraphQL but in a way not 100% compatible with the GraphQL-over-HTTP spec), or it can throw an error.

Maybe we could add a non-normative note to make this clearer? Or indicate that there may be more ways to talk GraphQL with a webserver than strictly defined in this spec?

Reply:
>I think it's okay for Automatic Persisted Queries to not be an official GraphQL request - the point of the specification is to define a common set of rules that means a compliant client from any vendor can speak to a compliant server from any vendor. APQ is not part of that flow, it's a specific behavior that the client/server can negotiate as an extension, separate from this specification.

Yep, that's fair. I just wanted to clarify that with our "automatic" PQ implementation, you don't have "a server that only allows persisted operations".

Re the second point, no strong feelings here (I think the spec is fine as given).